### DISCUSSION TITLE
Add support for JpegXL image format

### DISCUSSION THREAD

**Discussion by Gyanreyer** [withastro/roadmap#623] [Proposal]
### Body

# Summary

I would love to see support for the JpegXL format added to `astro:assets` as both an input and output image format.

# Background & Motivation

Given Apple's announcement that they will be supporting JpegXL in the next version of Safari/MacOS/iOS, the format will now be supported in a real browser without an experimental flag for the first time!

As such, I think it would make sense for Astro to add support for JpegXL assets so people can start using it.

It looks like Sharp already has support for jxl, but I'm uncertain about what sqoosh's level of support looks like so that could be a potential issue.

# Goals

- Add support for `"jxl"` as an input and output image format

**Comment by Chailotl** [comment]
I would like to be able to use JPEG XL as the output image format for `<Picture />` elements. In recent news, Chrome is adding back support for JPEG XL.

**Comment by delucis** [comment]
The default image service uses Sharp which supports JPEGXL so I think this may already be supported. Might just be a question of updating some types.

**Comment by jpc-ae** [comment]
Would love to see this as well. I don't have time to test right now, but I'd assume just adding jxl to `VALID_INPUT_FORMATS`, `VALID_SUPPORTED_FORMATS` and `VALID_OUTPUT_FORMATS` in `/packages/astro/src/assets/consts.ts` would be enough?

**Reply by Princesseuh** [reply]
No, Sharp does not ship with Jpegxl support. You need to compile libvips manually which is quite tedious. Until it's enabled by default in Sharp, I don't think I feel comfortable putting it an actual option. 

Note that if you do compile libvips manually with jpegxl enabled, you could use it right now in Astro

**Reply by jpc-ae** [reply]
Unfortunately, while I have a version of libvips compiled with libjxl, "The use of a globally-installed libvips is unsupported on Windows and on macOS when running Node.js under Rosetta." In other words, it seems like I would need to recompile sharp to use it in those environments. I can understand not wanting to deal with the tediousness of a custom version of sharp, but now that it's finally going (back) into browsers, we may see some traction on their side.

**Comment by jpc-ae** [comment]
For anyone wanting to get this working on Windows, the easiest way is to use Arch Linux inside Windows' WSL2. Here's a basic guide tutorial. Note that the `> ` at the beginning of the commands indicates a Windows prompt, while `$ ` indicates a Linux prompt. It should be obvious which is which while following the steps below.

## Setting up Arch Linux

1. Open [Windows Terminal](https://apps.microsoft.com/detail/9n0dx20hk701)
2. [Install WSL2](https://learn.microsoft.com/en-us/windows/wsl/install) by typing the following command and hitting `Enter`:

```
> wsl --install
```

3. [Install Arch Linux on WSL](https://wiki.archlinux.org/title/Install_Arch_Linux_on_WSL) using the following command:

```
> wsl --install archlinux`
```

4. Inside Arch Linux, install sudo and create a sudo user. `-Syu` will ensure all packages are up to date before installing; `--noconfirm` bypasses the 'are you sure [Y/n]' prompt; the `wheel` group where you put your user is the default admin group in Arch; Feel free to call your user something other than admin:

```
$ pacman -Syu sudo --noconfirm
$ echo "%wheel ALL=(ALL) ALL" > /etc/sudoers.d/wheel
$ useradd -m -G wheel -s /bin/bash admin
$ passwd admin
<input password>
```

5. Exit Arch Linux

```
$ exit
```

6. Switch the default login to your new user using the Windows wsl command (you can always switch it back to root if needed)

```
> wsl --manage archlinux --set-default-user admin
```

7. Start up Arch Linux again either in a new Windows Terminal tab or using the following command:

```
> wsl ~
```

8. Install the packages you need (insert your password when prompted). Feel free to replace nodejs with nvm if you need to manage different versions, and replace pnpm with npm, yarn or even corepack depending on your needs):

```
$ sudo pacman -Syu wget curl base-devel pkg-config git nodejs pnpm libvips libheif libjxl ffmpeg python --noconfirm
```

- `wget` and `curl` are for handling online or bleeding edge installations
- `base-devel` provides a variety of compilers including gcc which can compile sharp's C++17 code
- `pkg-config` is used by sharp to confirm your global libvips version
- `git`, `nodejs` and `pnpm` should be self-explanitory
- `libvips` provides a global copy for sharp to find, while `libheif` provides both HEIF and AVIF support and `libjxl` provides the all-important JPEG XL support
- `ffmpeg` assists with sharp's AVIF pipeline
- `python` is used for sharp's build scripts
- `--noconfirm` picks defaults for above confirmations. Note that this will cause ffmpeg to install with the older, more compatible jack2 instead of pipewire-jack, but in WSL there's not much of an issue either way. You can always leave this off if you're doing more with WSL and want pipewire.

If you really want to get rid of all warnings, add `imagemagick`, `openslide` and `poppler-glib` to the install list, but they're generally not needed for most. 

## Setting up Sharp

If you've already got a project set up on Windows, you can either access it through the mounted drive (`/mnt/c/...') or move the folder into your Linux environment. My preferred way is to create a separate virtual disk file that I [mount through wsl](https://learn.microsoft.com/en-us/windows/wsl/wsl2-mount-disk#mount-a-vhd-in-wsl) so the files are not tied to the distribution, but that's beyond the scope of this tutorial. The following is how to start a new project as a subfolder of your home drive by following [the Astro Setup tutorial](https://docs.astro.build/en/tutorial/1-setup/2/) and [adding sharp if you're using pnpm](https://docs.astro.build/en/reference/errors/missing-sharp/) like me.

1. Create a new Astro project:

```
$ pnpm create astro@latest
```

2. Follow the prompts, then `cd` into the new project:

```
$ cd astro-prj
```

3. Add [the npm dependencies required for sharp to build from source](https://sharp.pixelplumbing.com/install/#building-from-source):

```
$ pnpm add node-addon-api node-gyp
```

4. At this point you may get a `Warning` about build scripts. In that case run the following to approve them (at the prompt, tap `a` and `Enter`):

```
$ pnpm approve-builds
```

5. If you try adding an image to your pnpm-managed project and importing it into a `Picture` or `Image` element, you will see the `missing sharp` error mentioned earlier. Add it manually now:

```
$ pnpm add sharp
```

6. For some reason, the dual sharp dependencies caused the build script to not run automatically. Assuming you don't see any build script output, `cd` into the sharp folder and run it's build script manually:

```
$ cd node_modules/sharp
$ pnpm build
```

7. If you see `gyp info ok` you should be good to go. Don't forget to `cd` back out of the sharp folder before running any other commands.

```
$ cd ../../
```

And that's it! You can now try building with jxl image support.

## Tips

Processing a lot of formats can take time. You can save yourself some headaches by creating a constant with just the formats you need while using the astro dev server. For example:

```jsx
---
import { Picture } from 'astro:assets'
import ImageSrc from '../images/image.jpg'

const PictureFormats = import.meta.env.PROD ? ['jxl', 'avif', 'webp', 'jpg'] : ['jpg']
---
…
<Picture src={ImageSrc} formats={PictureFormats} alt='' />
…
```
